### DISCUSSION TITLE
Using c++ SDK in non MFC app, HOW???

### DISCUSSION THREAD

**Discussion by SiliconKiwi** [AxisCommunications/discussions#18] [Axis Windows SDKs] [answered]
MFC has been deprecated for many many years, but there is no examples of using this SDK outside of MFC, basically no info at all, I had to dive into the source code to get the CLSID, using CLSIDFromProgID( L"AxisMediaParserLib.AxisMediaParser", &clsid ), which is documented nowhere. Now I'm stuck trying to get media events (IAxisMediaParserEvents), seems to be no way to register a callback. Code uses AfxConnectionAdvise. What the hell does that do? Why is this so difficult?

**Comment by marbali8** [comment, marked as answer]
Hello!

Thank you for getting in touch with us. I will redirect this question to the appropriate team and update you as soon as possible.

**Reply by marbali8** [reply]
Hey. AMP SDK includes sample code of how to use AMP in .net with C#. Is it an option for you to use .net?

**Reply by SiliconKiwi** [reply]
No, its a Qt app. But I only want a frame event with data, should be easy? In the (also COM based) BlackMagic SDK its as easy as deckLinkInput_->SetCallback( &callbacks ); Where callbacks is simply an class derived from IDeckLinkInputCallback. Super easy.

**Reply by SiliconKiwi** [reply]
I have managed to use ATL to coble together a solution for now. Does require an unfortunate dependency on ATL

Something like:

```
#include "axismediaparserx64.tlh"

#include <iostream>
#include <atlbase.h>
#include <atlcom.h>
#include <comip.h>
///////////////////////////////////////////////////////////////////////
///
struct AxisEvents : 
	public CComObjectRootEx< CComMultiThreadModel >,
	public CComCoClass< AxisEvents, &CLSID_AxisMediaParser>,
    public IDispEventSimpleImpl< 234231341, AxisEvents, &DIID_IAxisMediaParserEvents > 
{
	//	
	BEGIN_SINK_MAP( AxisEvents )
		SINK_ENTRY_INFO( 234231341, DIID_IAxisMediaParserEvents, 0x67, &OnVideoSample, &onVideoType )
    END_SINK_MAP()

	static inline _ATL_FUNC_INFO onVideoType = { CC_STDCALL, VT_EMPTY, 6, {VT_I4, VT_I4, VT_I4, VT_I8, VT_I8, VT_VARIANT} };

    HRESULT __stdcall OnVideoSample(
        long CookieID,
        long SampleType,
        long SampleFlags,
        unsigned __int64 StartTime,
        unsigned __int64 StopTime,
        const _variant_t& SampleArray 
	);
};
```
Can be used like this:

```
	if ( !events_ ) {
		events_ = new AxisEvents;
	}

	if ( !media ) {

		CoCreateInstance( CLSID_AxisMediaParser, NULL, CLSCTX_ALL, __uuidof(IAxisMediaParser), (LPVOID*)&media );
		
		if ( media ) {
			HRESULT hr = media->QueryInterface(IID_IUnknown, (void**)&media_unknown); 
			if ( media_unknown ) {
				events_->DispEventAdvise( media_unknown );
			}
		}

	}
```

You need to dump a dummy ATL class somewhere global to initialize stuff...

```
class CDummyModule : public CAtlExeModuleT<CDummyModule> {};
CDummyModule DummyModule;
```

I would like to remove the ATL requirement, but diving into that old COM event, pre dual-interface gunk is draining my will to live, so will have to do for now.
